### PR TITLE
feat: implement knowledge v1beta2

### PR DESCRIPTION
**Pull request by cowan-macady**
BREAKING CHANGE: ENG-2778
- add knowledge v1beta2
- add objects v1beta2
- update helper
- remove KG config node
- change tests

### CODE REVIEW

File: examples/knowledge/cmd/read.go [hunk 1]
```diff
@@ -32,15 +32,20 @@ var readCmd = &cobra.Command{
 	Args:  cobra.ExactArgs(0),
 	Run: func(cmd *cobra.Command, args []string) {
 
-		path := "(n:Person)-[:OWNS]->(s:Store)"
-		conditions := "WHERE n.external_id = $external_id"
+		query := "MATCH (n:Person)-[r:CAN_SEE]->(a:Asset) WHERE n.id=$id"
 		params := map[string]*knowledgepb.InputParam{
 			"external_id": {
 				Value: &knowledgepb.InputParam_StringValue{StringValue: "1234"},
```
Comment (maaland) [inline examples/knowledge/cmd/read.go:38]: in the query you say `WHERE n.id = $id` but in the input params you provide `external_id`. You need to commit to one 😄

File: knowledge/example_test.go [hunk 2]
```diff
@@ -56,15 +57,20 @@ func ExampleClient_Read() {
 		_ = client.Close()
 	}()
 
-	path := "(n:Person)-[:HAS]->(s:Store)"
-	conditions := "WHERE n.external_id = $external_id"
+	query := "(n:Person)-[:HAS]->(s:Store) WHERE n.external_id = $external_id"
```
Comment (maaland) [inline knowledge/example_test.go:60]: you need to include `MATCH` at the start of the query

File: knowledge/example_test.go [hunk 3]
```diff
@@ -56,15 +57,20 @@ func ExampleClient_Read() {
...
 	params := map[string]*knowledgepb.InputParam{
 		"external_id": {
 			Value: &knowledgepb.InputParam_StringValue{StringValue: "1234"},
 		},
 	}
+	returns := []*knowledgepb.Return{
+		{
+			Variable:   "n",
+			Properties: []string{},
```
Comment (maaland) [inline knowledge/example_test.go:69]: this can just be unspecified/nil if you don't want to provide properties

File: knowledge/example_test.go [hunk 4]
```diff
@@ -74,8 +80,8 @@ func ExampleClient_Read() {
 	fmt.Println(json.Format(response))
 }
 
-// This example demonstrates how to use the Identity Knowledge Client to get a DigitalTwin by its id.
-func ExampleClient_GetDigitalTwinByID() {
+// This example demonstrates how to use the Identity Knowledge Client to get a Identity by its id.
```
Comment (maaland) [inline knowledge/example_test.go:83]:
```suggestion
// This example demonstrates how to use the Identity Knowledge Client to get an Identity by its id.
```

File: knowledge/example_test.go [hunk 5]
```diff
@@ -84,27 +90,14 @@ func ExampleClient_GetDigitalTwinByID() {
...
-
-// This example demonstrates how to use the Identity Knowledge Client to get a Resource by its id.
-func ExampleClient_GetResourceByID() {
-	client, err := knowledge.NewClient(context.Background())
-	if err != nil {
-		log.Fatalf("failed to create client %v", err)
+	returns := []*knowledgepb.Return{
+		{
+			Variable:   "n",
+			Properties: []string{},
```
Comment (maaland) [inline knowledge/example_test.go:96]: same here, can be nil

File: knowledge/example_test.go [hunk 6]
```diff
@@ -114,9 +107,9 @@ func ExampleClient_GetResourceByID() {
 	fmt.Println(json.Format(response))
 }
 
-// This example demonstrates how to use the Identity Knowledge Client to get a DigitalTwin
+// This example demonstrates how to use the Identity Knowledge Client to get a Identity
```
Comment (maaland) [inline knowledge/example_test.go:110]:
```suggestion
// This example demonstrates how to use the Identity Knowledge Client to get an Identity
```

File: knowledge/example_test.go [hunk 7]
```diff
@@ -125,34 +118,19 @@ func ExampleClient_GetDigitalTwinByIdentifier() {
...
-	})
-	if err != nil {
-		log.Fatalf("failed to invoke operation on IndyKite Client %v", err)
-	}
-	json := protojson.MarshalOptions{
-		Multiline: true,
+	returns := []*knowledgepb.Return{
+		{
+			Variable:   "n",
+			Properties: []string{},
```
Comment (maaland) [inline knowledge/example_test.go:124]: same here and other places, I won't mention all 😄

File: knowledge/knowledge_read.go [hunk 8]
```diff
@@ -23,195 +23,205 @@ import (
...
-	path := fmt.Sprintf("(:%s)", nodeType)
-	resp, err := c.client.IdentityKnowledge(ctx, &knowledgepb.IdentityKnowledgeRequest{
-		Path:      path,
-		Operation: knowledgepb.Operation_OPERATION_READ,
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = fmt.Sprintf(
+			"%s WHERE n.is_identity=%t",
+			query,
+			isIdentity,
```
Comment (maaland) [inline knowledge/knowledge_read.go:102]: all nodes that are identities will have the `DigitalTwin` label. So instead of this if, you can just change the query to `query := "MATCH (n:DigitalTwin)`

File: knowledge/knowledge_read.go [hunk 9]
```diff
@@ -23,195 +23,205 @@ import (
...
+		property.Type,
+		property.Type,
+	)
+	if isIdentity {
+		query = fmt.Sprintf(
+			"%s AND n.is_identity=%t",
+			query,
+			isIdentity,
+		)
+	}
```
Comment (maaland) [inline knowledge/knowledge_read.go:139]: same here, just use the `:DigitalTwin` label if you need to look for an identity

**Review by maaland — CHANGES_REQUESTED** [review]
Mostly good, just some comments.

I see you're replacing the old version with the new version. Can you first check with @yetizer / @bakkelim that that's okay for them? It might cause issues if they need a fix to something that comes in a later SDK version, but can't update yet because they're not ready to move to IKG API v2

File: examples/knowledge/cmd/read.go (hunk 1 above)
Comment (cowan-macady) [reply examples/knowledge/cmd/read.go]: updated

File: knowledge/example_test.go (hunk 2 above)
Comment (cowan-macady) [reply knowledge/example_test.go]: oops

File: knowledge/example_test.go (hunk 3 above)
Comment (cowan-macady) [reply knowledge/example_test.go]: updated

File: knowledge/example_test.go (hunk 4 above)
Comment (cowan-macady) [reply knowledge/example_test.go]: I can't believe it

File: knowledge/example_test.go (hunk 7 above)
Comment (cowan-macady) [reply knowledge/example_test.go]: updated

File: knowledge/knowledge_read.go (hunk 9 above)
Comment (cowan-macady) [reply knowledge/knowledge_read.go]: updated

File: knowledge/knowledge_read.go (hunk 8 above)
Comment (cowan-macady) [reply knowledge/knowledge_read.go]: updated

File: examples/knowledge/cmd/read.go [hunk 10]
```diff
@@ -32,15 +32,20 @@ var readCmd = &cobra.Command{
...
+		query := "MATCH (n:Person)-[r:CAN_SEE]->(a:Asset) WHERE n.external_id=$external_id"
 		params := map[string]*knowledgepb.InputParam{
 			"external_id": {
 				Value: &knowledgepb.InputParam_StringValue{StringValue: "1234"},
 			},
 		}
+		returns := []*knowledgepb.Return{
+			{
+				Variable:   "n",
+				Properties: []string{},
```
Comment (maaland) [inline examples/knowledge/cmd/read.go:44]: you can just omit the `properties` line as it will have the same effect

File: knowledge/knowledge_read.go [hunk 11]
```diff
@@ -23,190 +23,184 @@ import (
...
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)"
+	}
+	query = fmt.Sprintf(
+		"%s WHERE n.%s=$%s",
+		query,
+		property.Type,
+		property.Type,
+	)
```
Comment (maaland) [inline knowledge/knowledge_read.go:131]: This won't work anymore (it does right now, but only because we are temporarily backwards compatible with the old property model). You need to look for the property node
```
MATCH (n:Resource)-[:HAS]->(p:Property)
WHERE p.type = $type AND p.value = $value
```
and then the appropriate input params

File: knowledge/knowledge_read.go [hunk 12]
```diff
@@ -23,190 +23,184 @@ import (
...
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)"
+	}
+	query = fmt.Sprintf(
+		"%s WHERE n.%s=$%s",
+		query,
+		ID,
+		ID,
+	)
```
Comment (maaland) [inline knowledge/knowledge_read.go:173]: But this is fine, since the `id` is stored on the node, not as a separate property node. Same with `external_id`

File: knowledge/knowledge_read.go [hunk 13]
```diff
@@ -263,9 +254,8 @@ type Identifier struct {
 }
 
 const (
-	DigitalTwin = "DigitalTwin"
-	Resource    = "Resource"
-	ExternalID  = "external_id"
-	Type        = "type"
-	ID          = "id"
+	isIdentity = "is_identity"
```
Comment (maaland) [inline knowledge/knowledge_read.go:257]: don't think you need this const?

**Review by maaland — CHANGES_REQUESTED** [review]

File: knowledge/knowledge_read.go (hunk 11 above)
Comment (cowan-macady) [reply knowledge/knowledge_read.go]: updated

File: knowledge/knowledge_read.go (hunk 13 above)
Comment (cowan-macady) [reply knowledge/knowledge_read.go]: removed

**Comment by cowan-macady** [comment]
To be continued with modifications from objects v2

File: examples/knowledge/cmd/property.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+// propertyCmd represents the command for making a listNodesByProperty query with the Identity Knowledge API
+var propertyCmd = &cobra.Command{
+	Use:   "property",
+	Short: "Make a list query to the IndyKite Identity Knowledge API",
+	Long:  `Make a list query to the IndyKite Identity Knowledge API`,
+	Args:  cobra.ExactArgs(0),
+	Run: func(cmd *cobra.Command, args []string) {
+
+		query := "MATCH (n:Resource)-[:HAS]->(p:Property) WHERE p.type='last_name' and p.value='mushu'"
```
Comment (maaland) [inline examples/knowledge/cmd/property.go:37]: rather use the input params properly instead `p.type = $type and p.value = $value` and then provide `type` and `value` in the input params

File: knowledge/knowledge_read.go [hunk 15]
```diff
@@ -23,190 +23,188 @@ import (
...
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodes(ctx, Resource, opts...)
+	opts ...grpc.CallOption) ([]*knowledgeobjects.Node, error) {
+	return c.ListNodes(ctx, true, opts...)
 }
 
-// ListNodes is a helper function that lists all nodes by type, regardless of whether they are DigitalTwins
-// or Resources. The nodeType argument should be in PascalCase.
+// ListNodes is a helper function that lists all nodes by type, regardless of whether they are Identities
+// or not.
```
Comment (maaland) [inline knowledge/knowledge_read.go:85]: The functionality does not correspond with the comment. Why did you change the logic? Now the function lists all nodes that are either identities or not identities, depending on the arguments. Before, it listed all nodes of the provided nodeType argument, regardless of whether they were identities or not.

**Review by maaland — CHANGES_REQUESTED** [review]

File: knowledge/knowledge_read.go (hunk 15 above)
Comment (cowan-macady) [reply knowledge/knowledge_read.go]: updated

File: examples/knowledge/cmd/property.go (hunk 14 above)
Comment (cowan-macady) [reply examples/knowledge/cmd/property.go]: updated

File: knowledge/knowledge_read.go [hunk 16]
```diff
@@ -23,238 +23,226 @@ import (
...
+			Type: &objects.Value_IntegerValue{
 				IntegerValue: property.GetValue().GetIntegerValue()}}
 	case *objects.Value_StringValue:
-		params[property.Key] = &knowledgepb.InputParam{
-			Value: &knowledgepb.InputParam_StringValue{
+		params[property.Type] = &objects.Value{
+			Type: &objects.Value_StringValue{
 				StringValue: property.GetValue().GetStringValue()}}
 	default:
 		return nil, errors.New(codes.InvalidArgument, "only string or integer properties can be used for queries")
```
Comment (maaland) [inline knowledge/knowledge_read.go:139]: we support more types than just `string` or `integer` now:
```
                        *objectspb.Value_StringValue,
			*objectspb.Value_BoolValue,
			*objectspb.Value_IntegerValue,
			*objectspb.Value_DoubleValue,
			*objectspb.Value_TimeValue,
			*objectspb.Value_DurationValue,
			*objectspb.Value_ArrayValue:
```

**Review by maaland — APPROVED** [review]
Looks fine, just add more possible query values

File: knowledge/knowledge_read.go (hunk 16 above)
Comment (cowan-macady) [reply knowledge/knowledge_read.go]: added
